### PR TITLE
Increment 5

### PR DESCRIPTION
**Pull request by npavlov**

### CODE REVIEW

File: internal/agent/metrics/metrics.go [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+func (m *MetricService) UpdateMetrics() {
+	var memStats runtime.MemStats
+	runtime.ReadMemStats(&memStats)
+
+	m.Storage.UpdateGauge(types.Alloc, float64(memStats.Alloc))
+	m.Storage.UpdateGauge(types.BuckHashSys, float64(memStats.BuckHashSys))
+	m.Storage.UpdateGauge(types.Frees, float64(memStats.Frees))
+	m.Storage.UpdateGauge(types.GCCPUFraction, memStats.GCCPUFraction)
+	m.Storage.UpdateGauge(types.GCSys, float64(memStats.GCSys))
+	m.Storage.UpdateGauge(types.HeapAlloc, float64(memStats.HeapAlloc))
```
Comment (arttet) [inline internal/agent/metrics/metrics.go:38]: [CLARIFICATION] Please read the task again. Why did you convert everything to `float64`?

File: internal/types/handlers.go [hunk 2]
```diff
@@ -0,0 +1,9 @@
+package types
+
+import "net/http"
+
+type Handlers struct {
+	UpdateHandler   func(http.ResponseWriter, *http.Request)
```
Comment (arttet) [inline internal/types/handlers.go:6]: [CLARIFICATION] What did you want to do it here? It seems you can avoid it.

**Review by arttet — CHANGES_REQUESTED** [review]
- [ ]  Please integrate CodeCov https://about.codecov.io/blog/getting-started-with-code-coverage-for-golang/
- [x]  Please read and use the Conventional Commits specification https://www.conventionalcommits.org/en/v1.0.0/
- [ ] Please delete deprecated constraints in CI https://github.com/npavlov/go-metrics-service/blob/7cce4e5db4c2a87da8fbd0040efb889e36d4e094/.github/workflows/mertricstest.yml#L66

File: Makefile [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+clean:
+	rm -rf bin/
+
+run-server:
+	$(GO) run ./cmd/server/*.go
+
+run-client:
+	$(GO) run ./cmd/server/*.go
+
+lint:
```
[the remark this answers is not in the document]
Comment (npavlov) [reply Makefile:35]: phony added, please verify

File: Makefile [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+	golangci-lint run ./...
```
[the remark this answers is not in the document]
Comment (npavlov) [reply Makefile:36]: linter has been integrated into CI. check please

File: cmd/agent/flags.go [hunk 5]
```diff
@@ -0,0 +1,40 @@
+package main
+
+import (
+	"flag"
+	"fmt"
+	"github.com/caarlos0/env/v6"
+)
+
+var flagRunAddr string
```
[the remark this answers is not in the document]
Comment (npavlov) [reply cmd/agent/flags.go:9]: fixed

File: cmd/agent/main.go [hunk 6]
```diff
@@ -1,3 +1,70 @@
...
+	fmt.Printf("Enpoint address set as %s\n", flagRunAddr)
+	var service metrics.Service = metrics.NewMetricService(memStorage, "http://"+flagRunAddr)
+
+	fmt.Printf("Polling time time %d, reporint time %d\n", pollInterval, reportInterval)
+
+	// Create a context that will be canceled when a shutdown signal is received
+	ctx, cancel := context.WithCancel(context.Background())
+
+	// Channel to listen for OS signals
+	sigChan := make(chan os.Signal, 1)
```
[the remark this answers is not in the document]
Comment (npavlov) [reply cmd/agent/main.go:30]: moved into shutdown package

File: internal/types/handlers.go (hunk 2 above)
Comment (npavlov) [reply internal/types/handlers.go:6]: image

to use as an object and pass to the router, maybe there is a better way

File: internal/server/handlers/update/handler.go [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+		if err := updateMetric(ms, metricType, metricName, metricValue); err != nil {
+			http.Error(w, err.Error(), http.StatusBadRequest)
+			return
+		}
+
+		w.WriteHeader(http.StatusOK)
+	}
+}
+
+func updateMetric(ms storage.Repository, metricType types.MetricType, metricName types.MetricName, metricValue string) error {
```
[the remark this answers is not in the document]
Comment (npavlov) [reply internal/server/handlers/update/handler.go:33]: fixed

File: internal/types/handlers.go (hunk 2 above)
Comment (npavlov) [reply internal/types/handlers.go:6]: I've changed the structure, please review

File: internal/server/handlers/update/handler.go [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+	case types.Gauge:
+		value, err := strconv.ParseFloat(metricValue, 64)
+		if err != nil {
+			return errors.New(errInvalidGauge)
+		}
+		ms.UpdateGauge(metricName, value)
+	case types.Counter:
+		value, err := strconv.ParseInt(metricValue, 10, 64)
+		if err != nil {
+			return errors.New(errInvalidCounter)
```
[the remark this answers is not in the document]
Comment (npavlov) [reply internal/server/handlers/update/handler.go:44]: fixed

File: internal/server/handlers/render/handle.go [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+
+import (
+	"github.com/npavlov/go-metrics-service/internal/storage"
+	"github.com/npavlov/go-metrics-service/internal/types"
+	"html/template"
+	"net/http"
+)
+
+// Define the template for rendering the HTML page.
+const htmlTemplate = `
```
[the remark this answers is not in the document]
Comment (npavlov) [reply internal/server/handlers/render/handle.go]: fixed, using embed templates

File: internal/server/handlers/update/handler.go [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+	"strconv"
+)
+
+const (
+	errInvalidGauge   = "invalid gauge value"
+	errInvalidCounter = "invalid counter value"
+	errUnknownMetric  = "unknown metric type"
+)
+
+func GetUpdateHandler(ms storage.Repository) func(http.ResponseWriter, *http.Request) {
```
[the remark this answers is not in the document]
Comment (npavlov) [reply internal/server/handlers/update/handler.go]: created a new structure with handlers

File: internal/agent/metrics/metrics.go (hunk 1 above)
Comment (npavlov) [reply internal/agent/metrics/metrics.go]: A new structure Metrics has been created for such purposes

File: internal/server/handlers/handlers.go [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+
+func (mh *MetricHandler) Retrieve(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("Content-Type", "application/text")
+	metricType := types.MetricType(chi.URLParam(r, "metricType"))
+	metricName := types.MetricName(chi.URLParam(r, "metricName"))
+
+	switch metricType {
+	case types.Gauge:
+		if value, found := mh.st.GetGauge(metricName); found {
+			_, _ = w.Write([]byte(strconv.FormatFloat(value, 'f', -1, 64)))
```
Comment (arttet) [inline internal/server/handlers/handlers.go:59]: [CHANGE REQUEST] Please verify errors

File: internal/server/handlers/handlers.go [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+			return
+		}
+		http.Error(w, "unknown metric name", http.StatusNotFound)
+	default:
+		http.Error(w, "unknown metric type", http.StatusNotFound)
+	}
+}
+
+func (mh *MetricHandler) Render(w http.ResponseWriter, _ *http.Request) {
+	type MetricsPage struct {
```
Comment (arttet) [inline internal/server/handlers/handlers.go:75]:
```suggestion
	type metricsPage struct {
```

File: internal/server/handlers/handlers.go [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+		Gauges   map[types.MetricName]float64
+		Counters map[types.MetricName]int64
+	}
+
+	page := MetricsPage{
```
Comment (arttet) [inline internal/server/handlers/handlers.go:80]: [SUGGESTION] It looks like it should be in another layer.

File: internal/shutdown/shutdown.go [hunk 14]
```diff
@@ -0,0 +1,28 @@
+package shutdown
```
Comment (arttet) [inline internal/shutdown/shutdown.go:1]: [SUGGESTION] it looks like `utils`

File: internal/storage/storage.go [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+// GetCounter - retrieves the value of a counter
+func (ms *MemStorage) GetCounter(name types.MetricName) (int64, bool) {
+	ms.mu.RLock()
+	defer ms.mu.RUnlock()
+	value, exists := ms.counters[name]
+	return value, exists
+}
+
+// Generic function to clone a map with either int64 or float64 values
+func cloneMap[K comparable, V int64 | float64](original map[K]V) map[K]V {
```
Comment (arttet) [inline internal/storage/storage.go:69]:
```suggestion
func cloneMap[K comparable, V Number](original map[K]V) map[K]V {
```

https://stackoverflow.com/questions/67678331/how-can-i-write-a-generic-function-that-accepts-any-numerical-type

File: internal/types/metrictypes.go [hunk 16]
```diff
@@ -0,0 +1,49 @@
+package types
```
Comment (arttet) [inline internal/types/metrictypes.go:1]: [SUGGESTION] It should be `domain`

File: .github/workflows/golangci-lint.yml [hunk 17]
```diff
@@ -0,0 +1,26 @@
+name: golangci-lint
+on:
```
[the remark this answers is not in the document]
Comment (npavlov) [reply .github/workflows/golangci-lint.yml:2]: fixed

File: Makefile [hunk 18]
```diff
@@ -0,0 +1,48 @@
...
+# Define Go command, which can be overridden
+GO ?= go
+
+# Build the server binary from the Go source files in the cmd/server directory
+build-server:
+	$(GO) build -o bin/server ./cmd/server/*.go
+
+# Build the agent binary from the Go source files in the cmd/agent directory
+build-agent:
+	$(GO) build -o bin/agent ./cmd/agent/*.go
```
[the remark this answers is not in the document]
Comment (npavlov) [reply Makefile:12]: fixed

File: internal/storage/storage.go (hunk 15 above)
Comment (npavlov) [reply internal/storage/storage.go:69]: fixed

File: internal/types/metrictypes.go (hunk 16 above)
Comment (npavlov) [reply internal/types/metrictypes.go:1]: fixed

File: internal/shutdown/shutdown.go (hunk 14 above)
Comment (npavlov) [reply internal/shutdown/shutdown.go:1]: fixed

File: internal/server/handlers/handlers.go (hunk 12 above)
Comment (npavlov) [reply internal/server/handlers/handlers.go:75]: fixed

File: internal/server/handlers/handlers.go (hunk 13 above)
Comment (npavlov) [reply internal/server/handlers/handlers.go:80]: Created layer for retrieving templates

File: cmd/agent/main.go [hunk 19]
```diff
@@ -1,3 +1,37 @@
...
+	"github.com/npavlov/go-metrics-service/internal/shutdown"
+	"sync"
+)
+
+func main() {
+	cfg := NewConfigBuilder().
+		FromEnv().
+		FromFlags().Build()
+
+	st := stats.Stats{}
```
[the remark this answers is not in the document]
Comment (npavlov) [reply cmd/agent/main.go:17]: fixed

File: .github/workflows/golangci-lint.yml [hunk 20]
```diff
@@ -0,0 +1,26 @@
...
+  push:
+    branches:
+      - main
+      - master
```
[the remark this answers is not in the document]
Comment (npavlov) [reply .github/workflows/golangci-lint.yml:6]: fixed

**Pull request merged**